### DISCUSSION TITLE
Floating point exception

### DISCUSSION THREAD

**Discussion by aryanrahimieh** [galacticusorg/galacticus#416] [Q&A] [answered]
I changed this part of the xml file in order to use a file for the transfer function but I’m getting an error. I was wondering if I'm using it right.
I changed this part:
```
<!-- Power spectrum options -->
  <!-- Use the Bode et al. (2001) transfer function for thermal WDM -->
  <transferFunction value="bode2001">
    <epsilon value="0.359" />
    <eta value="3.810" />
    <nu value="1.100" />
    <!-- Bode2001 transfer function works by modifying a CDM transfer function - so feed it a CDM transfer function here -->
    <transferFunction value="eisensteinHu1999">
      <!-- Feed this transfer function a CDM particle - otherwise it will see the WDM particle defined above and complain that it
           can not compute WDM transfer functions -->
      <darkMatterParticle value="CDM" />
      <neutrinoNumberEffective value="3.046"/>
      <neutrinoMassSummed value="0.000"/>
    </transferFunction>
  </transferFunction>
```
into this:

```
<!-- Power spectrum options -->
  <!-- Use the Bode et al. (2001) transfer function for thermal WDM -->
  <transferFunction value="file">
    <fileName value="%DATASTATICPATH%/largeScaleStructure/transferFunctionWDM.hdf5"/>
  </transferFunction>
```
And this is the error that I'm getting: 

`Galacticus experienced a floating point exception - will try to flush data before exiting.`

**Comment by abensonca** [comment]
That change to the parameter file looks reasonable. Is there any additional output before the error message? If so, post it here - it might have some useful indications of possible causes. It's also possible that the problem is in your `transferFunctionWDM.hdf5` file - for example if there are wavenumbers at which the transfer function is zero I think this will cause problems (as the interpolation is done in $\log(k)$ vs. $\log(T)$.

You should be able to create the HDF5 file using any language - e.g. `h5py` in Python. If you're finding errors trying to read the file post them here.

**Comment by aryanrahimieh** [comment]
I checked the log10 of k and T and the vectors looks healthy. I also changed only one of the values of the transfer function and I received the same error. I created this hdf5 file with built-in tools in `hfpy` library. I was wondering how does the `transferFunctionMillenniumSimulation.hdf5` file was created originally.

**Reply by abensonca** [reply]
Can you share the hdf5 file and the parameter file you're using? I'll need to try actually running them to further diagnose the problem.

**Reply by aryanrahimieh** [reply]
Sure, it seems that I cannot send it here (the file type is not supported).

**Comment by abensonca** [comment, marked as answer]
I think the problem here is in the definition of what we mean by "transfer function" in this case. Galacticus defines the transfer function, $T(k)$, through:

$P(k) = T^2(k) P^\prime(k)$

where $P(k) is the power spectrum at late times and $P^\prime(k)$ is the primordial power spectrum.

In your case I think you are assuming:

$P(k) = T_\mathrm{IDM}^2(k) T^2_\mathrm{CDM}(k) P^\prime(k)$

and supplying in your file just $T_\mathrm{IDM}$. But, that means that Galacticus is attempting to evaluate:

$P(k) = T_\mathrm{IDM}^2(k) P^\prime(k)$

so it's missing the CDM part of the transfer function. 

In this case, you can correct the problem by telling it to apply your $T_\mathrm{IDM}$ on top of some other transfer function. For example, if you change the `transferFunction` section of your parameter file to:
```
  <transferFunction value="file">
    <fileName value="/home1/rahimieh/IDM_galacticus/Galacticus/datasets-master/static/largeScaleStructure/transferFunction_wdm_6p1kev_full.hdf5"/>
    <transferFunctionReference value="eisensteinHu1999">
      <neutrinoNumberEffective value="3.046"/>
      <neutrinoMassSummed value="0.000"/>
    </transferFunctionReference>
  </transferFunction>
```
then it will apply your transfer function on top of an `eisensteinHu1999` CDM transfer function. When I make this change your model runs correctly.

A few other points:

You should use a CDM particle (not thermal WDM) for this model as I believe you're capturing all DM particle effects in your transfer function and treating the particle as collisionless CDM after that:
```
<darkMatterParticle value="CDM">
```

Galacticus warns that the cosmological parameters set in the parameter file do not match those in your transfer function file. It will run anyway, but you most likely want to make these values consistent. If the values in your transfer function are correct then in the Galacticus parameter file you will need to set:
```
     <HubbleConstant value="73.0"/>
     <OmegaMatter value=" 0.25"/>
     <OmegaDarkEnergy value=" 0.75"/>
```

I would not use the the Barkana et al. (2001) method for the critical overdensity for collapse - this is specific to WDM. So change this to:
```
   <criticalOverdensity value="sphericalCollapseClsnlssMttrCsmlgclCnstnt">                                                                                                                                                               
```

**Comment by aryanrahimieh** [comment]
Thank you Andrew. This is really helpful. It worked!

**Reply by abensonca** [reply]
Happy to help!
